src/Native/Unix/System.Native/pal_signal.c

```diff
@@ -0,0 +1,287 @@
+static volatile CtrlCallback g_ctrlCallback = NULL; // Callback invoked for SIGINT/SIGQUIT
+static volatile SigChldCallback g_sigChldCallback = NULL; // Callback invoked for SIGCHLD
+static int g_signalPipe[2] = {-1, -1}; // Pipe used between signal handler and worker
+
+static struct sigaction* OrigActionFor(int sig)
+{
+    switch (sig)
+    {
+        case SIGINT:  return &g_origSigIntHandler;
+        case SIGQUIT: return &g_origSigQuitHandler;
+        case SIGCONT: return &g_origSigContHandler;
+        case SIGCHLD: return &g_origSigChldHandler;
+    }
+
+    assert(false);
+    return NULL;
+}
+
+static void SignalHandler(int sig, siginfo_t* siginfo, void* context)
+{
+    if (sig == SIGCONT || sig == SIGCHLD)
+    {
+        // SIGCONT will be sent when we're resumed after suspension, at which point
+        // we need to set the terminal back up.  Similarly, SIGCHLD will be sent after
+        // a child process completes, and that child could have left things in a bad state,
+        // so we similarly need to reinitialize.
+        ReinitializeConsole();
+    }
+
+    // Signal handler for signals where we want our background thread to do the real processing.
+    // It simply writes the signal code to a pipe that's read by the thread.
+    if (sig == SIGQUIT || sig == SIGINT || sig == SIGCHLD)
+    {
+        // Write the signal code to the pipe
+        uint8_t signalCodeByte = (uint8_t)sig;
+        ssize_t writtenBytes;
+        while ((writtenBytes = write(g_signalPipe[1], &signalCodeByte, 1)) < 0 && errno == EINTR);
+
+        if (writtenBytes != 1)
+        {
+            abort(); // fatal error
+        }
+    }
+
+    // Delegate to any saved handler we may have
+    if (sig == SIGCONT)
+    {
+        struct sigaction* origHandler = OrigActionFor(sig);
+        if (origHandler->sa_sigaction != NULL &&
+            (void*)origHandler->sa_sigaction != (void*)SIG_DFL &&
+            (void*)origHandler->sa_sigaction != (void*)SIG_IGN)
+        {
+            origHandler->sa_sigaction(sig, siginfo, context);
+        }
+    }
+}
+
+void SystemNative_ResumeSigChld()
+{
+    struct sigaction* origHandler = OrigActionFor(SIGCHLD);
+
+    if ((void*)origHandler->sa_sigaction == (void*)SIG_IGN)
+    {
+        // When the original disposition is SIG_IGN, children that terminated did not become zombies.
+        // Since we overwrote the disposition, we are now responsible for reaping those processes.
+        pid_t pid;
+        do
+        {
+            int status;
+            while ((pid = waitpid(-1, &status, WNOHANG)) < 0 && errno == EINTR);
+        } while (pid > 0);
+    }
+    else if ((void*)origHandler->sa_sigaction == (void*)SIG_DFL)
+    {
+        // do nothing
+    }
+    else if (origHandler->sa_sigaction != NULL)
+    {
+        // TODO?: We are passing a NULL siginfo and context, do we need to try and do better?
```

Comment:
TODO?

Reply:
Oh, I see.  This seems problematic, no?  That we're not passing the original information down, on the original thread, etc.?  Why can't we do this delegation to the original as part of the actual signal handler rather than in the asynchronous handler, e.g. have the async handler delegate to the original and then queue the work to the separate thread?

Reply:
> This seems problematic, no? That we're not passing the original information down, on the original thread, etc.?

Since SIGCHLD can be merged, you can not count on getting this info for each child. So it's less problematic than it seems at first.

I think we must properly handle SIG_DLF case, and should handle SIG_IGN.
If there is a custom handler in place, I think we always end up making some assumptions on its behavior. If we call it from the signal thread, we are assuming it won't reap our children. I think we can document that assumption and move this to the signal thread.